src/twisted/conch/newsfragments/9412.bugfix

```diff
@@ -0,0 +1 @@
+If the command run by twisted.conch.endpoints.SSHCommandClientEndpoint exits because of a delivered signal, the client protocol's connectionLost is now called with a ProcessTerminated exception instead of a ConnectionDone exception.
```

Comment:
That code never ran because `struct.unpack` would fail with an exception instead of parsing a well-formed `exit-signal` message:

```python
def request_exit_signal(self, data):
    """
    When the server sends the command's exit status, record it for later
    delivery to the protocol.
    @param data: The network-order four byte representation of the exit
        signal of the command.
    @type data: L{bytes}
    """
    (signal,) = unpack('>L', data)  # raises: struct.error: unpack requires a string argument of length 4
    self._reason = ProcessTerminated(None, signal, None)
```

You can see this on trunk by running:

```
>>> from twisted.conch.endpoints import _CommandChannel
>>> packet = '\x00\x00\x00\x04TERM\x01\x00\x00\x00\x07message\x00\x00\x00\x05en-US'
>>> _CommandChannel(None, None, None, None).request_exit_signal(packet)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/twisted/conch/endpoints.py", line 229, in request_exit_signal
    (signal,) = unpack('>L', data)
struct.error: unpack requires a string argument of length 4
```

[`log.callWithLogger`](https://twistedmatrix.com/documents/18.4.0/api/twisted.python.log.html#callWithLogger) suppresses the exception inside `SSHTransportBase.dispatchMessage`:

https://github.com/twisted/twisted/blob/9d6dbd7e0584bd10a502b4c99045dbd5e794abc5/src/twisted/conch/ssh/transport.py#L774-L775

Here's a partial traceback courtesy of a `pdb` breakpoint inside `_CommandChannel.request_exit_signal`:

```
  /home/mrw/src/python/twisted/src/twisted/internet/endpoints.py(125)dataReceived()
-> return self._wrappedProtocol.dataReceived(data)
  /home/mrw/src/python/twisted/src/twisted/conch/ssh/transport.py(750)dataReceived()
-> self.dispatchMessage(messageNum, packet[1:])
  /home/mrw/src/python/twisted/src/twisted/conch/ssh/transport.py(775)dispatchMessage()
-> messageNum, payload)
  /home/mrw/src/python/twisted/src/twisted/python/log.py(103)callWithLogger()
-> return callWithContext({"system": lp}, func, *args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/log.py(86)callWithContext()
-> return context.call({ILogContext: newCtx}, func, *args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/context.py(122)callWithContext()
-> return self.currentContext().callWithContext(ctx, func, *args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/context.py(85)callWithContext()
-> return func(*args,**kw)
  /home/mrw/src/python/twisted/src/twisted/conch/ssh/service.py(45)packetReceived()
-> return f(packet)
  /home/mrw/src/python/twisted/src/twisted/conch/ssh/connection.py(322)ssh_CHANNEL_REQUEST()
-> channel.requestReceived, requestType, rest[1:])
  /home/mrw/src/python/twisted/src/twisted/internet/defer.py(150)maybeDeferred()
-> result = f(*args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/log.py(103)callWithLogger()
-> return callWithContext({"system": lp}, func, *args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/log.py(86)callWithContext()
-> return context.call({ILogContext: newCtx}, func, *args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/context.py(122)callWithContext()
-> return self.currentContext().callWithContext(ctx, func, *args, **kw)
  /home/mrw/src/python/twisted/src/twisted/python/context.py(85)callWithContext()
-> return func(*args,**kw)
  /home/mrw/src/python/twisted/src/twisted/conch/ssh/channel.py(162)requestReceived()
-> return f(data)
> /home/mrw/src/python/twisted/src/twisted/conch/endpoints.py(230)request_exit_signal()
-> (signal,) = unpack('>L', data)
```

Because the exception is suppressed, the code related to `connectionLost` continues to run without setting the `_reason` attribute on the `_CommandChannel` instance.  Eventually `_CommandChannel.closed` runs, finds that its `_reason` attribute is `None`, and calls back the underlying protocol with a default `ConnectionDone` instance:

https://github.com/twisted/twisted/blob/9d6dbd7e0584bd10a502b4c99045dbd5e794abc5/src/twisted/conch/endpoints.py#L233-L243

In other words, the exception raised by `struct.unpack` prevents `_CommandChannel.request_exit_signal` from setting its `_reason` to `ProcessTerminated`, so calls the protocol's `connectionLost` method with `ConnectionDone` instead.

This PR has `request_exit_signal` actually parse the signal number out of the packet and thus set `_reason` to `ProcessTerminated`.  The packet contains additional information that cannot be included in `ProcessTerminated`, like whether or not a core dump occurred. I chose to log so that information so users could see it and the tests could assert the packet was fully parsed.